python/cornserve/services/resource_manager/manager.py

```diff
@@ -242,6 +242,116 @@ async def wait_for_online(rank: int) -> None:
+                        node_id=gpu.node,
+                        global_rank=gpu.global_rank,
+                        local_rank=gpu.local_rank,
+                    )
+                    for gpu in resources
+                ]
+                update_resource_req = task_manager_pb2.UpdateResourcesRequest(task_manager_id=task_state.id, gpus=gpus)
+                response = await task_state.stub.UpdateResources(update_resource_req)
+                if response.status != common_pb2.Status.STATUS_OK:
+                    raise RuntimeError(
+                        f"UpdateResources gRPC error when trying to scale up task manager: {response}",
+                    )
+                # Add the new GPUs to the task state resources
+                async with self.task_states_lock:
+                    if task_state.resources is None:
+                        task_state.resources = []
+                    task_state.resources.extend(resources)
+            except Exception as e:
+                async with self.task_states_lock:
+                    for gpu in resources:
+                        gpu.free()
+                logger.exception("Failed to scale up task %s by %d GPUs: %s", task, num_gpus, e)
+                raise RuntimeError(f"Failed to scale up task {task} by {num_gpus} GPUs: {e}") from e
+
+    @tracer.start_as_current_span("ResourceManager.scale_down_unit_task")
+    async def scale_down_unit_task(self, task: UnitTask, num_gpus: int) -> None:
+        """Scale down a unit task by removing GPUs from the task manager."""
+        assert num_gpus > 0, "Number of GPUs to scale down must be positive"
+        logger.info("Scaling down unit task %s by %d GPUs", task, num_gpus)
+
+        span = trace.get_current_span()
+        span.set_attribute("resource_manager.scale_down_unit_task.task", str(task))
+        span.set_attribute("resource_manager.scale_down_unit_task.num_gpus", num_gpus)
+
+        # Find the task manager state for this task
+        task_state = None
+        async with self.task_states_lock:
+            for state in self.task_states.values():
+                if state.deployment and state.deployment.task == task:
+                    task_state = state
+
+            if task_state is None:
+                logger.info("Task %s is not running, returning immediately", task)
+                return
+
+        assert task_state.stub is not None, "Task manager stub is not initialized"
+        # check there are enough GPUs to scale down
+        if task_state.resources is None or len(task_state.resources) < num_gpus:
+            raise RuntimeError(
+                f"{task} has only "
+                f"{len(task_state.resources) if task_state.resources else 0} GPUs, "
+                f"cannot scale down by {num_gpus}"
+            )
+        async with task_state.lock:
+            try:
+                gpus_to_remove = task_state.resources[:num_gpus]
```

Comment:
Eventually, this should be part of the resource allocator. For now, probably leave a TODO comment and create an issue. We will likely have preferences for which GPU to deallocate from the task manager. For instance, when a unit task has 4 GPUs in one node and 2 GPUs in another node, from which node should we deallocate? This requires a lot of information to do well, and will be part of the autoscaling feature.